parlai/mturk/core/react_server/dev/app.jsx

```diff
@@ -0,0 +1,668 @@
+/*
+ * Copyright (c) 2017-present, Facebook, Inc.
+ * All rights reserved.
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+import React from 'react';
+import ReactDOM from 'react-dom';
+import {FormControl, Button} from 'react-bootstrap';
+import CustomComponents from './components/custom.jsx';
+import SocketHandler from './components/socket_handler.jsx';
+import {MTurkSubmitForm, allDoneCallback} from './components/mturk_submit_form.jsx';
+import 'fetch';
+
+/* ================= Utility functions ================= */
+
+// Determine if the browser is a mobile phone
+function isMobile() {
+  return /Android|webOS|iPhone|iPad|iPod|BlackBerry|IEMobile|Opera Mini/i.test(
+    navigator.userAgent);
+}
+
+// Sends a request to get the hit_config
+function getHitConfig(callback_function) {
+  $.ajax({
+    url: '/get_hit_config',
+    timeout: 3000 // in milliseconds
+  }).retry({times: 10, timeout: 3000}).then(
+    function(data) {
+      if (callback_function) {
+        callback_function(data);
+      }
+    }
+  );
+}
+
+// Sees if the current browser supports WebSockets
+function doesSupportWebsockets() {
+  return !((bowser.msie && bowser.version < 10) ||
+           (bowser.firefox && bowser.version < 11) ||
+           (bowser.chrome && bowser.version < 16) ||
+           (bowser.safari && bowser.version < 7) ||
+           (bowser.opera && bowser.version < 12.1));
+}
+
+class ChatMessage extends React.Component {
+  render() {
+    let float_loc = 'left';
+    let alert_class = 'alert-warning';
+    if (this.props.is_self) {
+      float_loc = 'right';
+      alert_class = 'alert-info';
+    }
+    return (
+      <div className={"row"} style={{'marginLeft': '0', 'marginRight': '0'}}>
+        <div
+          className={"alert " + alert_class} role="alert"
+          style={{'float': float_loc, 'display': 'table'}}>
+          <span style={{'fontSize': '16px'}}>
+            <b>{this.props.agent_id}</b>: {this.props.message}
+          </span>
+        </div>
+      </div>
+    );
+  }
+}
+
+class MessageList extends React.Component {
+  makeMessages() {
+    let agent_id = this.props.agent_id;
+    let messages = this.props.messages;
+    return messages.map(
+      m => <XChatMessage
+        key={m.message_id}
+        is_self={m.id == agent_id}
+        agent_id={m.id}
```

Comment:
I should add a note here - this could be a message from the agent (outgoing messages) or it could be a message from anyone else, so this is taking the `'id'` field out of the act dict. Figuring out whether it appears as a self message or from someone else is determined the line previous

Reply:
Caught this a moment ago - that was the real bug. Fixed locally and I'll commit it with my first round of improvements now that the core is in